File: src/lexer.rs
use std::collections::HashMap;
use crate::error::RuntimeError;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    // Single-character tokens
    LeftParen, RightParen, LeftBrace, RightBrace, LeftBracket, RightBracket,
    Comma, Dot, Minus, Plus, Semicolon, Divide, Multiply,

    // One or two character tokens
    Not, NotEqual,
    Assign, Equal,
    Greater, GreaterEqual,
    Less, LessEqual,
    And, Or,

    // Literals
    Identifier, String, Number, Char,

    // Keywords
    Print, Else, False, For, Function, If, Return, True, While, Import, From,
    IntType, FloatType, CharType, StringType, List, Map,
    Loadlib, Getproc, Freelib, Callext,
    Assert, Var, Nil,

    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub lexeme: String,
    pub literal: Option<LiteralValue>,
    pub line: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub enum LiteralValue {
    String(String),
    Number(f64),
    Integer(i64),
    Boolean(bool),
    Char(char),
    Nil,
}

impl Token {
    pub fn new(token_type: TokenType, lexeme: String, literal: Option<LiteralValue>, line: usize) -> Self {
        Self {
            token_type,
            lexeme,
            literal,
            line,
        }
    }
}

pub struct Lexer {
    source: Vec<char>,
    filename: String,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl Lexer {
    pub fn new(source: &str, filename: &str) -> Self {
        Self {
            source: source.chars().collect(),
            filename: filename.to_string(),
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn scan_tokens(&mut self) -> Result<Vec<Token>, RuntimeError> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token()?;
        }

        self.tokens.push(Token::new(TokenType::Eof, String::new(), None, self.line));
        Ok(self.tokens.clone())
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> char {
        let ch = self.source[self.current];
        self.current += 1;
        ch
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            '\0'
        } else {
            self.source[self.current]
        }
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            '\0'
        } else {
            self.source[self.current + 1]
        }
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() || self.source[self.current] != expected {
            false
        } else {
            self.current += 1;
            true
        }
    }

    fn add_token(&mut self, token_type: TokenType, literal: Option<LiteralValue>) {
        let text: String = self.source[self.start..self.current].iter().collect();
        self.tokens.push(Token::new(token_type, text, literal, self.line));
    }

    fn string(&mut self) -> Result<(), RuntimeError> {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            return Err(RuntimeError::new(
                "Unterminated string.".to_string(),
                Some(self.line),
                &self.filename,
            ));
        }

        self.advance(); // The closing "
        let value: String = self.source[self.start + 1..self.current - 1].iter().collect();
        self.add_token(TokenType::String, Some(LiteralValue::String(value)));
        Ok(())
    }

    fn number(&mut self) {
        let mut is_float = false;
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        if self.peek() == '.' && self.peek_next().is_ascii_digit() {
            is_float = true;
            self.advance(); // Consume the "."
            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        let text: String = self.source[self.start..self.current].iter().collect();
        if is_float {
            let value: f64 = text.parse().unwrap();
            self.add_token(TokenType::Number, Some(LiteralValue::Number(value)));
        } else {
            let value: i64 = text.parse().unwrap();
            self.add_token(TokenType::Number, Some(LiteralValue::Integer(value)));
        }
    }

    fn identifier(&mut self) {
        while self.peek().is_alphanumeric() || self.peek() == '_' {
            self.advance();
        }

        let text: String = self.source[self.start..self.current].iter().collect();
        let token_type = self.get_keyword(&text).unwrap_or(TokenType::Identifier);
        
        let literal = match token_type {
            TokenType::True => Some(LiteralValue::Boolean(true)),
            TokenType::False => Some(LiteralValue::Boolean(false)),
            TokenType::Nil => Some(LiteralValue::Nil),
            _ => None,
        };

        self.add_token(token_type, literal);
    }

    fn get_keyword(&self, text: &str) -> Option<TokenType> {
        let mut keywords = HashMap::new();
        keywords.insert("print", TokenType::Print);
        keywords.insert("if", TokenType::If);
        keywords.insert("else", TokenType::Else);
        keywords.insert("while", TokenType::While);
        keywords.insert("for", TokenType::For);
        keywords.insert("function", TokenType::Function);
        keywords.insert("return", TokenType::Return);
        keywords.insert("true", TokenType::True);
        keywords.insert("false", TokenType::False);
        keywords.insert("import", TokenType::Import);
        keywords.insert("from", TokenType::From);
        keywords.insert("int", TokenType::IntType);
        keywords.insert("float", TokenType::FloatType);
        keywords.insert("char", TokenType::CharType);
        keywords.insert("string", TokenType::StringType);
        keywords.insert("list", TokenType::List);
        keywords.insert("map", TokenType::Map);
        keywords.insert("loadlib", TokenType::Loadlib);
        keywords.insert("getproc", TokenType::Getproc);
        keywords.insert("freelib", TokenType::Freelib);
        keywords.insert("callext", TokenType::Callext);
        keywords.insert("assert", TokenType::Assert);
        keywords.insert("var", TokenType::Var);
        keywords.insert("nil", TokenType::Nil);

        keywords.get(text).cloned()
    }

    fn scan_token(&mut self) -> Result<(), RuntimeError> {
        let c = self.advance();

        match c {
            ' ' | '\r' | '\t' => {}, // Ignore whitespace
            '\n' => self.line += 1,
            '(' => self.add_token(TokenType::LeftParen, None),
            ')' => self.add_token(TokenType::RightParen, None),
            '{' => self.add_token(TokenType::LeftBrace, None),
            '}' => self.add_token(TokenType::RightBrace, None),
            '[' => self.add_token(TokenType::LeftBracket, None),
            ']' => self.add_token(TokenType::RightBracket, None),
            ',' => self.add_token(TokenType::Comma, None),
            '.' => self.add_token(TokenType::Dot, None),
            '-' => self.add_token(TokenType::Minus, None),
            '+' => self.add_token(TokenType::Plus, None),
            ';' => self.add_token(TokenType::Semicolon, None),
            '*' => self.add_token(TokenType::Multiply, None),
            '!' => {
                let token_type = if self.match_char('=') {
                    TokenType::NotEqual
                } else {
                    TokenType::Not
                };
                self.add_token(token_type, None);
            },
            '=' => {
                let token_type = if self.match_char('=') {
                    TokenType::Equal
                } else {
                    TokenType::Assign
                };
                self.add_token(token_type, None);
            },
            '<' => {
                let token_type = if self.match_char('=') {
                    TokenType::LessEqual
                } else {
                    TokenType::Less
                };
                self.add_token(token_type, None);
            },
            '>' => {
                let token_type = if self.match_char('=') {
                    TokenType::GreaterEqual
                } else {
                    TokenType::Greater
                };
                self.add_token(token_type, None);
            },
            '|' => {
                if self.match_char('|') {
                    self.add_token(TokenType::Or, None);
                } else {
                    return Err(RuntimeError::new(
                        "Unexpected character: |".to_string(),
                        Some(self.line),
                        &self.filename,
                    ));
                }
            },
            '/' => {
                if self.match_char('/') {
                    // A comment goes until the end of the line
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else {
                    self.add_token(TokenType::Divide, None);
                }
            },
            '&' => {
                if self.match_char('&') {
                    self.add_token(TokenType::And, None);
                } else {
                    return Err(RuntimeError::new(
                        "Unexpected character: &".to_string(),
                        Some(self.line),
                        &self.filename,
                    ));
                }
            },
            '"' => self.string()?,
            '\'' => {
                let char_val = self.advance();
                if self.advance() != '\'' {
                    return Err(RuntimeError::new(
                        "Unterminated character literal.".to_string(),
                        Some(self.line),
                        &self.filename,
                    ));
                }
                self.add_token(TokenType::Char, Some(LiteralValue::Char(char_val)));
            },
            _ => {
                if c.is_ascii_digit() {
                    self.number();
                } else if c.is_alphabetic() || c == '_' {
                    self.identifier();
                } else {
                    return Err(RuntimeError::new(
                        format!("Unexpected character: {}", c),
                        Some(self.line),
                        &self.filename,
                    ));
                }
            }
        }

        Ok(())
    }
}

File: src/c/lexer.c
#include "mini_script.h"
#include <ctype.h>
#include <string.h>

/* Keywords table */
static const struct {
  const char *keyword;
  TokenType type;
} keywords[] = {{"and", AND},
                {"assert", ASSERT},
                {"char", CHAR_TYPE},
                {"else", ELSE},
                {"false", FALSE},
                {"float", FLOAT_TYPE},
                {"for", FOR},
                {"from", FROM},
                {"function", FUNCTION},
                {"if", IF},
                {"import", IMPORT},
                {"int", INT_TYPE},
                {"list", LIST},
                {"map", MAP},
                {"nil", NIL},
                {"or", OR},
                {"print", PRINT},
                {"return", RETURN},
                {"string", STRING_TYPE},
                {"true", TRUE},
                {"var", VAR},
                {"while", WHILE},
                {"loadlib", LOADLIB},
                {"getproc", GETPROC},
                {"freelib", FREELIB},
                {"callext", CALLEXT},
                {NULL, EOF_TOKEN}};

static TokenType get_keyword_type(const char *text) {
  for (int i = 0; keywords[i].keyword != NULL; i++) {
    if (strcmp(text, keywords[i].keyword) == 0) {
      return keywords[i].type;
    }
  }
  return IDENTIFIER;
}

static bool is_digit(char c) { return c >= '0' && c <= '9'; }

static bool is_alpha(char c) {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

static bool is_alphanumeric(char c) { return is_alpha(c) || is_digit(c); }

static void add_token(Lexer *lexer, TokenType type, LiteralValue *literal) {
  if (lexer->token_count >= lexer->token_capacity) {
    lexer->token_capacity =
        lexer->token_capacity == 0 ? 8 : lexer->token_capacity * 2;
    lexer->tokens =
        realloc(lexer->tokens, lexer->token_capacity * sizeof(Token));
  }

  size_t length = lexer->current - lexer->start;
  char *lexeme = malloc(length + 1);
  strncpy(lexeme, lexer->source + lexer->start, length);
  lexeme[length] = '\0';

  lexer->tokens[lexer->token_count].type = type;
  lexer->tokens[lexer->token_count].lexeme = lexeme;
  lexer->tokens[lexer->token_count].literal = literal;
  lexer->tokens[lexer->token_count].line = lexer->line;
  lexer->token_count++;
}

static char advance(Lexer *lexer) { return lexer->source[lexer->current++]; }

static bool match(Lexer *lexer, char expected) {
  if (lexer->source[lexer->current] == '\0')
    return false;
  if (lexer->source[lexer->current] != expected)
    return false;
  lexer->current++;
  return true;
}

static char peek(Lexer *lexer) { return lexer->source[lexer->current]; }

static char peek_next(Lexer *lexer) {
  if (lexer->source[lexer->current] == '\0')
    return '\0';
  return lexer->source[lexer->current + 1];
}

static void scan_string(Lexer *lexer) {
  while (peek(lexer) != '"' && peek(lexer) != '\0') {
    if (peek(lexer) == '\n')
      lexer->line++;
    advance(lexer);
  }

  if (peek(lexer) == '\0') {
    // Unterminated string - just return
    return;
  }

  advance(lexer); // The closing "

  // Trim the surrounding quotes
  size_t length = lexer->current - lexer->start - 2;
  char *value = malloc(length + 1);
  strncpy(value, lexer->source + lexer->start + 1, length);
  value[length] = '\0';

  LiteralValue *literal = literal_new(LITERAL_STRING);
  literal->value.string = value;
  /* Lexer owns raw string until parser duplicates; mark ownership */
  literal->owns_string = true;

  add_token(lexer, STRING, literal);
}

static void scan_number(Lexer *lexer) {
  while (is_digit(peek(lexer)))
    advance(lexer);

  // Look for fractional part
  if (peek(lexer) == '.' && is_digit(peek_next(lexer))) {
    advance(lexer); // Consume the "."
    while (is_digit(peek(lexer)))
      advance(lexer);

    size_t length = lexer->current - lexer->start;
    char *text = malloc(length + 1);
    strncpy(text, lexer->source + lexer->start, length);
    text[length] = '\0';

    LiteralValue *literal = literal_new(LITERAL_NUMBER);
    literal->value.number = atof(text);
    free(text);

    add_token(lexer, NUMBER, literal);
  } else {
    size_t length = lexer->current - lexer->start;
    char *text = malloc(length + 1);
    strncpy(text, lexer->source + lexer->start, length);
    text[length] = '\0';

    LiteralValue *literal = literal_new(LITERAL_INTEGER);
    literal->value.integer = atol(text);
    free(text);

    add_token(lexer, NUMBER, literal);
  }
}

static void scan_identifier(Lexer *lexer) {
  while (is_alphanumeric(peek(lexer)))
    advance(lexer);

  size_t length = lexer->current - lexer->start;
  char *text = malloc(length + 1);
  strncpy(text, lexer->source + lexer->start, length);
  text[length] = '\0';

  TokenType type = get_keyword_type(text);

  LiteralValue *literal = NULL;
  if (type == TRUE) {
    literal = literal_new(LITERAL_BOOLEAN);
    literal->value.boolean = true;
  } else if (type == FALSE) {
    literal = literal_new(LITERAL_BOOLEAN);
    literal->value.boolean = false;
  } else if (type == NIL) {
    literal = literal_new(LITERAL_NIL);
  }

  add_token(lexer, type, literal);
  free(text);
}

static void scan_token(Lexer *lexer) {
  char c = advance(lexer);

  switch (c) {
  case '(':
    add_token(lexer, LEFT_PAREN, NULL);
    break;
  case ')':
    add_token(lexer, RIGHT_PAREN, NULL);
    break;
  case '{':
    add_token(lexer, LEFT_BRACE, NULL);
    break;
  case '}':
    add_token(lexer, RIGHT_BRACE, NULL);
    break;
  case '[':
    add_token(lexer, LEFT_BRACKET, NULL);
    break;
  case ']':
    add_token(lexer, RIGHT_BRACKET, NULL);
    break;
  case ',':
    add_token(lexer, COMMA, NULL);
    break;
  case '.':
    add_token(lexer, DOT, NULL);
    break;
  case '-':
    add_token(lexer, MINUS, NULL);
    break;
  case '+':
    add_token(lexer, PLUS, NULL);
    break;
  case ';':
    add_token(lexer, SEMICOLON, NULL);
    break;
  case '*':
    add_token(lexer, MULTIPLY, NULL);
    break;
  case '/':
    if (match(lexer, '/')) {
      // A comment goes until the end of the line
      while (peek(lexer) != '\n' && peek(lexer) != '\0')
        advance(lexer);
    } else {
      add_token(lexer, DIVIDE, NULL);
    }
    break;
  case '!':
    add_token(lexer, match(lexer, '=') ? NOT_EQUAL : NOT, NULL);
    break;
  case '=':
    add_token(lexer, match(lexer, '=') ? EQUAL : ASSIGN, NULL);
    break;
  case '<':
    add_token(lexer, match(lexer, '=') ? LESS_EQUAL : LESS, NULL);
    break;
  case '>':
    add_token(lexer, match(lexer, '=') ? GREATER_EQUAL : GREATER, NULL);
    break;
  case '&':
    if (match(lexer, '&')) {
      add_token(lexer, AND, NULL);
    } else {
      // Single & is not supported, could be an error
      add_token(lexer, EOF_TOKEN, NULL); // Treat as error for now
    }
    break;
  case '|':
    if (match(lexer, '|')) {
      add_token(lexer, OR, NULL);
    } else {
      // Single | is not supported, could be an error
      add_token(lexer, EOF_TOKEN, NULL); // Treat as error for now
    }
    break;
  case ' ':
  case '\r':
  case '\t':
    // Ignore whitespace
    break;
  case '\n':
    lexer->line++;
    break;
  case '"':
    scan_string(lexer);
    break;
  case '\'':
    // Character literal
    if (peek(lexer) != '\0' && peek_next(lexer) == '\'') {
      char ch = advance(lexer);
      advance(lexer); // closing '

      LiteralValue *literal = literal_new(LITERAL_CHAR);
      literal->value.character = ch;
      add_token(lexer, CHAR, literal);
    }
    break;
  default:
    if (is_digit(c)) {
      scan_number(lexer);
    } else if (is_alpha(c)) {
      scan_identifier(lexer);
    } else {
      // Unexpected character - ignore
    }
    break;
  }
}

Lexer *lexer_new(const char *source) {
  Lexer *lexer = malloc(sizeof(Lexer));
  lexer->source = source;
  lexer->start = 0;
  lexer->current = 0;
  lexer->line = 1;
  lexer->tokens = NULL;
  lexer->token_count = 0;
  lexer->token_capacity = 0;
  return lexer;
}

void lexer_free(Lexer *lexer) {
  if (lexer) {
    for (size_t i = 0; i < lexer->token_count; i++) {
      free(lexer->tokens[i].lexeme);
      if (lexer->tokens[i].literal) {
        literal_free(lexer->tokens[i].literal);
      }
    }
    free(lexer->tokens);
    free(lexer);
  }
}

void lexer_scan_tokens(Lexer *lexer) {
  while (lexer->source[lexer->current] != '\0') {
    lexer->start = lexer->current;
    scan_token(lexer);
  }

  lexer->start = lexer->current;
  add_token(lexer, EOF_TOKEN, NULL);
}

/* Literal value functions */
LiteralValue *literal_new(LiteralType type) {
  LiteralValue *literal = malloc(sizeof(LiteralValue));
  literal->type = type;
  memset(&literal->value, 0, sizeof(literal->value));
  literal->owns_string = false;
  return literal;
}

void literal_free(LiteralValue *literal) {
  if (literal) {
    if (literal->type == LITERAL_STRING && literal->value.string &&
        literal->owns_string) {
      free(literal->value.string);
    }
    free(literal);
  }
}
